src/autopas/selectors/tuningStrategy/GaussianProcess.h

```diff
@@ -17,36 +21,122 @@ namespace autopas {
+     */
+    Hyperparameters()
+        : mean(0.),
+          theta(1.),
+          dimScales(Eigen::VectorXd::Ones(1)),
+          covMatInv(Eigen::MatrixXd::Ones(1, 1)),
+          weights(Eigen::VectorXd::Ones(1)) {}
+
+    /**
+     * Contructor
+     * @param mean prior mean of Gaussian process
+     * @param theta prior variance
+     * @param dimScales scale for each input dimension
+     */
+    Hyperparameters(double mean, double theta, const Eigen::VectorXd &dimScales)
+        : mean(mean), theta(theta), dimScales(dimScales) {}
+
+    /**
+     * Precalculate matrices needed for predictions
+     * @param sigma assumed noise
+     * @param inputs evidence input
+     * @param outputs evidence output
+     */
+    void precalculate(double sigma, const std::vector<Vector> &inputs, const Eigen::VectorXd &outputs) {
+      size_t size = outputs.size();
+      // mean of output shifted to zero
+      Eigen::VectorXd outputCentered = outputs - mean * Eigen::VectorXd::Ones(size);
+
+      Eigen::MatrixXd covMat(size, size);
+      // calculate covariance matrix
+      for (size_t i = 0; i < size; ++i) {
+        covMat(i, i) = kernel(inputs[i], inputs[i], theta, dimScales) + sigma;
+        for (size_t j = i + 1; j < size; ++j) {
+          covMat(i, j) = covMat(j, i) = kernel(inputs[i], inputs[j], theta, dimScales);
+        }
+      }
+
+      // cholesky decomposition
+      Eigen::LLT<Eigen::MatrixXd> llt = covMat.llt();
+      Eigen::MatrixXd l = llt.matrixL();
+
+      // precalculate inverse of covMat and weights for predictions
+      covMatInv = llt.solve(Eigen::MatrixXd::Identity(size, size));
+      weights = covMatInv * outputCentered;
+
+      // likelihood of evidence given parameters
+      score = std::exp(-0.5 * outputCentered.dot(weights)) / l.diagonal().prod();
+
+      if (std::isnan(score)) {
+        // error score calculation failed
+        utils::ExceptionHandler::exception("GaussianProcess: invalid score ", score);
+      }
+    }
+  };
+
  public:
   /**
    * Constructor
-   * @param theta prior variance
-   * @param dimScale scale of each dimension before applying kernel
+   * @param dims number of input dimensions
    * @param sigma fixed noise
+   * @param rngRef reference to rng
    */
-  GaussianProcess(double theta, std::vector<double> dimScale, double sigma)
-      : _inputs(),
-        _outputs(),
-        _theta(theta),
-        _dimScale(Eigen::Map<Eigen::VectorXd>(dimScale.data(), dimScale.size())),
-        _sigma(sigma),
-        _covMat(),
-        _covMatInv(),
-        _weights() {}
+  GaussianProcess(size_t dims, double sigma, Random &rngRef)
+      : _inputs(), _outputs(), _dims(dims), _sigma(sigma), _hypers(), _rng(rngRef) {
+    updateHyperparameters();
+  }
 
   /**
    * Discard all evidence.
    */
   void clear() {
-    // As long is _input is empty the matrices ar consided uninitialized.
-    // So no further actions needed.
     _inputs.clear();
+    updateHyperparameters();
```

Comment:
this is done to also reset the hyperparameters right? Document why you need to call this here.

Reply:
Apparently, this is not needed. If (_input.size()==0) predictions skip usage of hyperparameters.